python/sglang/srt/layers/quantization/deep_gemm.py

```diff
@@ -0,0 +1,304 @@
+import logging
+import os
+from contextlib import contextmanager
+from enum import IntEnum, auto
+from typing import Tuple
+
+import torch
+from tqdm.contrib.concurrent import thread_map
+
+from sglang.srt.utils import get_bool_env_var, get_device_sm, get_int_env_var, is_cuda
+
+ENABLE_JIT_DEEPGEMM = False
+ENABLE_JIT_DEEPGEMM_BMM = False
+ENABLE_JIT_DEEPGEMM_PRECOMPILE = get_bool_env_var("SGL_JIT_DEEPGEMM_PRECOMPILE", "true")
+ENABLE_JIT_DEEPGEMM_LOG_BUILD = get_bool_env_var(
+    "SGL_JIT_DEEPGEMM_LOG_BUILD", "false" if ENABLE_JIT_DEEPGEMM_PRECOMPILE else "true"
+)
+COMPILE_WORKERS = get_int_env_var("SGL_JIT_DEEPGEMM_COMPILE_WORKERS", 4)
+
+if is_cuda():
+    import deep_gemm
+    from deep_gemm import get_num_sms
+    from deep_gemm.jit_kernels.gemm import get_best_configs
+    from deep_gemm.jit_kernels.gemm import includes as deep_gemm_includes
+    from deep_gemm.jit_kernels.gemm import template as deep_gemm_gemm_template
+    from deep_gemm.jit_kernels.m_grouped_gemm import (
+        template as deep_gemm_grouped_gemm_template,
+    )
+    from deep_gemm.jit_kernels.tuner import jit_tuner
+
+    sm_version = get_device_sm()
+    if sm_version == 90:
+        if get_bool_env_var("SGL_ENABLE_JIT_DEEPGEMM", default="false"):
+            ENABLE_JIT_DEEPGEMM = True
+        if get_bool_env_var("SGL_ENABLE_JIT_DEEPGEMM_BMM", default="false"):
+            ENABLE_JIT_DEEPGEMM_BMM = True
+
+logger = logging.getLogger(__name__)
+
+INITIALIZATION_DICT = {}
+PRE_COMPILE_M_LIST = list(range(1, 1024 * 64 + 1))
+
+# Force redirect deep_gemm cache_dir
+os.environ["DG_CACHE_DIR"] = os.getenv(
+    "SGL_DG_CACHE_DIR", os.path.expanduser("~") + "/.cache/deep_gemm"
+)
+
+
+class DeepGemmKernelType(IntEnum):
+    GROUPED_GEMM_NT_F8F8BF16_MASKED = auto()
+    GROUPED_GEMM_NT_F8F8BF16_CONTIG = auto()
+    GEMM_NT_F8F8BF16 = auto()
+
+
+FUNCTION_NAME_DICT = {
+    DeepGemmKernelType.GROUPED_GEMM_NT_F8F8BF16_MASKED: "m_grouped_gemm_fp8_fp8_bf16_nt_masked",
+    DeepGemmKernelType.GROUPED_GEMM_NT_F8F8BF16_CONTIG: "m_grouped_gemm_fp8_fp8_bf16_nt_contiguous",
+    DeepGemmKernelType.GEMM_NT_F8F8BF16: "gemm_fp8_fp8_bf16_nt",
+}
+
+
+def compile_grouped_gemm_nt_f8f8bf16_masked_one(n, k, num_groups, config) -> None:
+    # Auto-tuning with compilation
+    global deep_gemm_includes, deep_gemm_grouped_gemm_template
+    _, block_m, block_n, num_stages, tma_multicast_config, smem_config = config
+    _ = jit_tuner.compile_and_tune(
```

Comment:
Can we do this using multiprocessing? And can we use multiple GPU devices simultaneously?

Reply:
Multithreading by `tqdm` is enough here because the main cost is on nvcc subprocess. I am going to let only first rank of a node do the compilation.

Reply:
Will it be faster if we use all ranks?

Reply:
Offline discussion with @zhyncs : Compilation with all ranks in one node cannnot get significantly speed up and may be harmful if we launch too many nvcc subprocess. Keep 4 threads is enough